functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,143 @@
+
+
+    true_graph = get_isolated_graphmodule(true_fn, operands, {})
+    false_graph = get_isolated_graphmodule(false_fn, operands, {})
+
+    # There are probably better ways - I know that create_arg has some self incrementing name
+    # magic to it, but since we explicitly have to get the name for register_module,
+    # I was not sure how to do that. This kinda simulates it.
+    next_name = None
+    i = 0
+    while not next_name:
+        candidate = f"true_graph_{i}"
+        if hasattr(proxy_mode.tracer.root, candidate):
+            i += 1
+        else:
+            next_name = candidate
+
+    random_slug = ''.join(random.choices(string.digits, k=5))
+    true_name = next_name
+    false_name = f"false_graph_{i}"
+
+    proxy_mode.tracer.root.register_module(true_name, true_graph)
+    proxy_mode.tracer.root.register_module(false_name, false_graph)
+
+    with no_dispatch():
+        # This is not amazing.
+        # However, if we have nested operators that have a call_function
+        # in their graph that is not a torch op (ex: see conditional below, nested cond)
+        # we cannot get metadata for it from just looking at out vars.
+        # The reason is that an operation on the output of such an op is not
+        # evalauted as a torch.Tensor.
+        # So we execute the real true and false fn here and compare metadata
+        # inp_ops = [o for o in operands]
+        true_result = true_graph(operands)
+        false_result = false_graph(operands)
+        def recursive_compare_same(a, b):
+            assert(type(a) == type(b))
+            if isinstance(a, torch.Tensor):
+                assert(a.dtype == b.dtype)
+                assert(a.size() == b.size())
+                assert(a.stride() == b.stride())
+                assert(a.device == b.device)
+            elif isinstance(a, (list, tuple)):
+                assert(len(a) == len(b))
+                for i in range(0, len(a)):
+                    recursive_compare_same(a[i], b[i])
+
+        recursive_compare_same(true_result, false_result)
+
+    args = (pred, true_graph, false_graph, operands)
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    proxy_res = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, kwargs,
+                                            name="conditional")
+    return proxy_res
+
+
+def cond_dense(pred, true_fn, false_fn, *operands):
+    mode = torch._C._get_torch_dispatch_mode()
+    assert (mode is None), "Mode should never be enabled for CPU key"
+    if pred:
+        return true_fn(*operands)
+    else:
+        return false_fn(*operands)
+
+
+def cond_autograd(pred, true_fn, false_fn, *operands):
+    # TODO: support autograd
+    flat_operands, _ = tree_flatten([true_fn, false_fn] + [operands])
+    assert all([not f.requires_grad for f in flat_operands
+                if isinstance(f, torch.Tensor)])
```

Comment:
A further improvement would be to test `torch.is_grad_enabled()` which will enable you to trace when `no_grad()` is set, even if the inputs require grad.

Reply:
There is a subtle soundness issue, in fact, with `requires_grad` testing. Suppose that `true_fn` or `false_fn` closes over a tensor that `requires_grad`. The "composite" semantics of conditional (that is, if we got rid of all of this dispatch business) would demand that we start tracking gradients in the inside of `true_fn`. This would be quite difficult to do correctly without pre-tracing `true_fn` and `false_fn` in order to determine the free variables and then setup autograd correctly. Luckily executorch will not have to worry about this for now, but handling this correctly in its full generality with training is going to be quite a project. It might be best if we force tracing the branches *before* we even start handling the dispatcher.

Reply:
Yea - we are punting the autograd impl on the assumption that the torch.cond API will stay the same. Are you fine with us kicking the generality can down the road for a bit?

Reply:
yes